File: scala/problem19.scala
/*
https://projecteuler.net/problem=19

Counting Sundays
*/

import java.time.DayOfWeek
import java.time.LocalDate

object problem19 {
  private def zeller(y: Int, m: Int, d: Int): Int = {
    var year = y
    var month = m
    if (month < 3) {
      month += 12
      year -= 1
    }

    val k = year % 100
    val j = year / 100
    (d + (13 * (month + 1)) / 5 + k + k / 4 + j / 4 + 5 * j) % 7
  }

  def baselineAnswer: Long = {
    val count =
      (for {
        year <- 1901 to 2000
        month <- 1 to 12
      } yield LocalDate.of(year, month, 1).getDayOfWeek)
        .count(_ == DayOfWeek.SUNDAY)

    count.toLong
  }

  def optimizedAnswer: Long = {
    var count = 0
    var year = 1901
    while (year <= 2000) {
      var month = 1
      while (month <= 12) {
        if (zeller(year, month, 1) == 1) {
          count += 1
        }
        month += 1
      }
      year += 1
    }
    count.toLong
  }

  case class Result(value: Long, timeNs: Long)

  def answer: Long = {
    optimizedAnswer
  }

  def eulerBaseline(): Result = {
    val start = System.nanoTime()
    val value = baselineAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def euler(): Result = {
    val start = System.nanoTime()
    val value = optimizedAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def main(args: Array[String]): Unit = {
    println("Project Euler Problem 19")

    val baseline = eulerBaseline()
    val optimized = euler()

    println(s"Result (baseline): ${baseline.value}")
    println(s"Time (baseline): ${baseline.timeNs} ns")
    println(s"Result (optimized): ${optimized.value}")
    println(s"Time (optimized): ${optimized.timeNs} ns")

    if (optimized.timeNs > 0) {
      val speedup = baseline.timeNs.toDouble / optimized.timeNs.toDouble
      println(s"Speedup: ${speedup}x")
    }
  }
}

File: cpp/problem19.cpp
/*
https://projecteuler.net/problem=19

Counting Sundays

How many Sundays fell on the first of the month during the twentieth century
(1 Jan 1901 to 31 Dec 2000)?
*/

#include <chrono>
#include <iostream>
using namespace std;

struct result {
    long long value;
    long long time_ns;
};

bool is_leap(int y) {
    if (y % 400 == 0) return true;
    if (y % 100 == 0) return false;
    return y % 4 == 0;
}

int days_in_month(int y, int m) {
    static const int days[] = {31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
    if (m == 2) {
        return days[1] + (is_leap(y) ? 1 : 0);
    }
    return days[m - 1];
}

long long evaluate_baseline() {
    int day_of_week = 1;
    long long count = 0;

    for (int year = 1900; year <= 2000; ++year) {
        for (int month = 1; month <= 12; ++month) {
            if (year >= 1901 && day_of_week == 0) {
                ++count;
            }
            day_of_week = (day_of_week + days_in_month(year, month)) % 7;
        }
    }

    return count;
}

int day_of_week_sakamoto(int y, int m, int d) {
    static int t[] = {0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4};
    if (m < 3) y -= 1;
    return (y + y / 4 - y / 100 + y / 400 + t[m - 1] + d) % 7;
}

long long evaluate_optimized() {
    long long count = 0;

    for (int y = 1901; y <= 2000; ++y) {
        for (int m = 1; m <= 12; ++m) {
            if (day_of_week_sakamoto(y, m, 1) == 0) {
                ++count;
            }
        }
    }

    return count;
}

result euler() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_optimized();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

result euler_baseline() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_baseline();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

int main() {
    std::cout << "Project Euler Problem 19" << std::endl;

    result baseline = euler_baseline();
    result optimized = euler();

    std::cout << "Result (baseline): " << baseline.value << std::endl;
    std::cout << "Time (baseline): " << baseline.time_ns << " ns" << std::endl;
    std::cout << "Result (optimized): " << optimized.value << std::endl;
    std::cout << "Time (optimized): " << optimized.time_ns << " ns" << std::endl;

    if (optimized.time_ns > 0) {
        double speedup = static_cast<double>(baseline.time_ns) / static_cast<double>(optimized.time_ns);
        std::cout << "Speedup: " << speedup << "x" << std::endl;
    }

    return 0;
}
